File: examples/typescript/code-lookup.ts
/**
 * Demonstrates Normatia Building Codes API usage with raw fetch.
 */

const DEFAULT_API_URL = "https://api.normatia.com";

function getClientConfig(): { apiKey: string; baseUrl: string } {
  const apiKey = process.env.NORMATIA_API_KEY;
  if (!apiKey) {
    console.error("Error: NORMATIA_API_KEY is not set.");
    console.error("Set it first: export NORMATIA_API_KEY='sk-normatia-xxxxx'");
    process.exit(1);
  }

  return {
    apiKey,
    baseUrl: (process.env.NORMATIA_API_URL ?? DEFAULT_API_URL).replace(/\/+$/, ""),
  };
}

function formatOutput(value: unknown): string {
  if (typeof value === "string") {
    return value;
  }
  return JSON.stringify(value, null, 2);
}

function withQuery(path: string, params: Record<string, string>): string {
  const query = new URLSearchParams(params);
  return `${path}?${query.toString()}`;
}

async function fetchJson(
  config: { apiKey: string; baseUrl: string },
  path: string,
): Promise<unknown> {
  const url = `${config.baseUrl}${path}`;

  let response: Response;
  try {
    response = await fetch(url, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${config.apiKey}`,
        Accept: "application/json",
      },
    });
  } catch (error) {
    throw new Error(`Network error while calling ${url}: ${String(error)}`);
  }

  const raw = await response.text();
  let payload: unknown = null;
  if (raw) {
    try {
      payload = JSON.parse(raw);
    } catch {
      payload = raw;
    }
  }

  if (!response.ok) {
    throw new Error(`Request failed (${response.status} ${response.statusText}) for ${url}\n${formatOutput(payload)}`);
  }

  return payload;
}

async function runCall(config: { apiKey: string; baseUrl: string }, title: string, path: string): Promise<void> {
  console.log(`\n== ${title} ==`);
  const payload = await fetchJson(config, path);
  console.log(formatOutput(payload));
}

async function main(): Promise<void> {
  const config = getClientConfig();

  await runCall(
    config,
    "1) Search codes by text: ahorro energetico",
    withQuery("/api/v1/codes/search", {
      q: "ahorro energetico",
      normative_scope: "national",
      page: "1",
      page_size: "5",
    }),
  );

  await runCall(config, "2) Code detail: cte-db-he", "/api/v1/codes/cte-db-he");

  await runCall(config, "3) List versions: cte-db-he", "/api/v1/codes/cte-db-he/versions");

  await runCall(
    config,
    "4) Version detail: cte-db-he 2022",
    "/api/v1/codes/cte-db-he/versions/2022",
  );
}

main().catch((error) => {
  console.error("\nCode lookup example failed.");
  console.error(error instanceof Error ? error.message : String(error));
  process.exit(1);
});

File: examples/python/code_lookup.py
#!/usr/bin/env python3
"""Demonstrate `search_codes` only using Normatia's search endpoint."""

from __future__ import annotations

import asyncio
import os
import sys
from typing import Any

import httpx

DEFAULT_API_URL = "https://api.normatia.com"


def get_client_config() -> tuple[str, dict[str, str]]:
    api_key = os.getenv("NORMATIA_API_KEY")
    if not api_key:
        raise RuntimeError(
            "Missing NORMATIA_API_KEY environment variable. "
            "Set it to your bearer key, for example: sk-normatia-xxxxx"
        )

    base_url = os.getenv("NORMATIA_API_URL", DEFAULT_API_URL).rstrip("/")
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Accept": "application/json",
    }
    return base_url, headers


def print_code_search(payload: dict[str, Any]) -> None:
    items = payload.get("items")
    if not isinstance(items, list):
        items = []

    total = payload.get("total")

    print("Code Search Results")
    print("===================")
    print(f"query: eficiencia energética")
    print(f"total: {total if total is not None else len(items)}")

    if not items:
        print("No codes found.")
        return

    for index, item in enumerate(items[:10], start=1):
        if not isinstance(item, dict):
            continue
        print(f"{index}. slug: {item.get('slug')}")
        print(f"   title: {item.get('title')}")
        print(f"   normative_scope: {item.get('normative_scope')}")


async def main() -> None:
    base_url, headers = get_client_config()

    try:
        async with httpx.AsyncClient(base_url=base_url, headers=headers, timeout=20.0) as client:
            search_response = await client.get(
                "/api/v1/codes/search",
                params={"q": "eficiencia energética"},
            )
            search_response.raise_for_status()
            search_payload = search_response.json()
            if not isinstance(search_payload, dict):
                raise RuntimeError("Unexpected response shape for code search.")
            print_code_search(search_payload)

    except httpx.HTTPStatusError as exc:
        print("\nRequest failed with a non-success HTTP status.")
        print(f"Status: {exc.response.status_code}")
        print(f"URL: {exc.request.url}")
        print(f"Body: {exc.response.text}")
        sys.exit(1)
    except httpx.RequestError as exc:
        print("\nNetwork error while calling Normatia API.")
        print(f"Error: {exc}")
        sys.exit(1)
    except RuntimeError as exc:
        print(f"\nConfiguration or response error: {exc}")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())
